### ISSUE TITLE
[QUESTION] I would like to fix the transition matrix upon running. Can I have some help pointing to what update steps need to be commented?

### ISSUE THREAD

**Issue by gajarajv** [jmschrei/pomegranate#1092]
**Describe the bug**
A clear and concise description of what the bug is, including what you were expecting to happen and what actually happened. Please report the version of pomegranate that you are using and the operating system. Also, please make sure that you have upgraded to the latest version of pomegranate before submitting the bug report.

**To Reproduce**
Please provide a snippet of code that can reproduce this error. It is much easier for us to track down bugs and fix them if we have an example script that fails until we're successful.

**Response time**
Although I will likely respond during weekdays if I am not on vacation, I am not likely to be able to merge PRs or write code until the weekend.

**Comment by gajarajv** [comment]
Specifically, when running baum welch for the fit method, I'd like to iterate through any given sample without updating parameter updates for the transition edges

**Comment by jmschrei** [comment]
If you pass in `inertia=1.0` to your HMM you should be able to freeze the transition matrix while keeping the emissions unfrozen.

**Comment by gajarajv** [comment]
Thanks, I am trying to populate the SparseHMM object using sparse (CSR) matrix and am unsure if I am using the newer functionality correctly. Right now I have several steps to input each of the values individually as such:

```python 
dists = [Categorical([row]) for row in E.toarray()]
ends = E.max(axis=1).data
ends /= np.sum(ends)
P0 /= np.sum(P0)

rows, cols = T.nonzero()
values = T.data

# Get corresponding distributions with transition prob
index_value_pairs = [(dists[row], dists[col], T[row, col]) for row, col in zip(rows, cols)]

model = SparseHMM(dists,edges =index_value_pairs, starts=P0,ends = ends, max_iter=1, verbose=True, inertia=1.0)

```
I wasn't able to find anything detailing this scenario on the docs but does this follow your design?

**Comment by jmschrei** [comment]
I think that `ends` should be derived from `T` not `E` right? Not sure where you're getting the first call for `P0` but, assuming it's right, I think what you have here is right.

**Comment by gajarajv** [comment]
Err, right. Now that I'm looking at it more carefully I am setting them as the stationary dist of the transition matrix, `T`. 

However I noticed that for some cases problems where I do not initialize the `ends` variable, the EM algorithm breaks down and produces nan solutions. Any clue as to why the setting of the `ends` var is very important? 

I haven't seen BW implementations that rely on the ending states, so I'm curious on the inspiration for this design.

**Comment by jmschrei** [comment]
I don't think BW here is dependent on the end states but having it can add constraints that make optimization easier. It's hard to diagnose why one might observe the behavior you describe without knowing more about your data or model.

**Comment by gajarajv** [comment]
From your understanding of your optimizer, can you provide some potential sources from where nans can come from in the code? I am trying to solve for a overcomplete HMM's emission matrix. Investigating, it looks like that some of the sources of the errors might be coming from way we initialize the problem. When the # of states and output possibilities are more reasonable (10 states, 5 observations) it looks like the solver can work well. However if we bring the number of states to 1000, with 20 possible observations (under genericly created conditions), I find that the solver output to look like this:
![image](https://github.com/jmschrei/pomegranate/assets/153683783/3950dbbd-c035-40e5-9945-124872a9e227)
Where the prob values in the solver are all NaN. 

For more context, here are the relative sizes of each of the objects:
![image](https://github.com/jmschrei/pomegranate/assets/153683783/91606486-c461-461a-8e37-ac4baa3ac67b)

I figure I am dividing by zero somewhere in the code where one of the forward backward computes runs into an inf. Is there a way of changing scaling or something along those lines to reduce the liklihood of errors like this?

**Comment by gajarajv** [comment]
Upon closer inspection with a debugger, it appears the problem is the result of integer underflow during the forward backward algorithm when the emission probabilities are also sparse.

**Comment by jmschrei** [comment]
Can you elaborate a little bit more? Where are you getting integer underflow? The probabilities should all be floats, right?

**Comment by gajarajv** [comment]
I'll try and get you a script that can verifiably replicate this problem sometime this week

**Comment by gajarajv** [comment]
Totally forgot to update this ticket but I solved this issue by the following modifications to the forward and backwards:
1. Initialization of f:
Use torch.finfo(torch.float64).min to initialize f to the minimum finite value representable by float64 rather than -inf to avoid issues in subsequent calculations.

2. Log-Sum-Exp Calculation:
Subtract alpha before the exponential operation (torch.exp(p - alpha)) and clip the result of the exponential operation to avoid zeros.

3. Zero Clamping in z:
Clamp z before taking the logarithm to ensure that torch.log(z) does not attempt to compute the log of zero.
Squeezing alpha:

4. Corrected the handling of alpha to remove unnecessary dimensions before adding it back into the forward variable calculation.
These changes should help stabilize the computation by managing extreme values and preventing -inf and NaN issues in the log-space calculations

```python
import torch
def forward(self, X=None, emissions=None, priors=None):

    emissions = self._check_inputs(X, emissions, priors)
    n, l, _ = emissions.shape

    f = torch.full((l, n, self.n_distributions), torch.finfo(torch.float64).min, dtype=torch.float64, 
                   device=self.device)
    f[0] = self.starts + emissions[:, 0]

    for i in range(1, l):
        p = f[i-1, :, self._edge_idx_starts]
        p += self._edge_log_probs.expand(n, -1)

        alpha = torch.max(p, dim=1, keepdims=True).values
        p = p - alpha  # Stabilized values before exp to prevent underflow
        p = torch.exp(p).clamp(min=1e-10)  # Prevents exp(0) leading to exact 0s

        z = torch.zeros_like(f[i])
        z.scatter_add_(1, self._edge_idx_ends.expand(n, -1), p)

        z = z.clamp(min=1e-10)  # Prevents log(0)
        f[i] = alpha.squeeze(1) + torch.log(z) + emissions[:, i]  # Corrected alpha handling

    f = f.permute(1, 0, 2)
    return f
```

**Comment by gajarajv** [comment]
```python   
def backward(self, X=None, emissions=None, priors=None):
    emissions = self._check_inputs(X, emissions, priors)
    n, l, _ = emissions.shape

    b = torch.full((l, n, self.n_distributions), torch.finfo(torch.float64).min, dtype=torch.float64,
                   device=self.device)
    b[-1] = self.ends + emissions[:, -1]

    for i in range(l-2, -1, -1):
        p = b[i+1, :, self._edge_idx_ends]
        p += emissions[:, i+1]
        p += self._edge_log_probs.expand(n, -1)

        alpha = torch.max(p, dim=1, keepdims=True).values
        p = p - alpha  # Stabilized values before exp to prevent underflow
        p = torch.exp(p).clamp(min=1e-10)  # Prevents exp(0) leading to exact 0s

        z = torch.zeros_like(b[i])
        z.scatter_add_(1, self._edge_idx_starts.expand(n, -1), p)

        z = z.clamp(min=1e-10)  # Prevents log(0)
        b[i] = alpha.squeeze(1) + torch.log(z)  # Corrected alpha handling

    b = b.permute(1, 0, 2)
    return b

**Comment by gajarajv** [comment]
If you think this is worth a PR, I can set that up

**Comment by jmschrei** [comment]
Sorry the delay in response. I'm not sure that I agree with the clamping to a small value because it prevents true -neginfs, right?
